File: app/csharp-ms/src/FinancialAccounting/FinancialAccounting.Domain/Entities/Account.cs
namespace FinancialAccounting.Domain.Entities;

/// <summary>
/// 勘定科目エンティティ
/// </summary>
public class Account
{
    public int AccountId { get; set; }
    public string AccountCode { get; set; } = string.Empty;
    public string AccountName { get; set; } = string.Empty;
    public string? AccountNameKana { get; set; }
    public string AccountType { get; set; } = string.Empty;  // 資産, 負債, 純資産, 収益, 費用
    public bool IsSummaryAccount { get; set; }
    public string BsPlType { get; set; } = string.Empty;     // B: 貸借対照表, P: 損益計算書
    public string TransactionElementType { get; set; } = string.Empty;
    public string? ExpenseType { get; set; }
    public int DisplayOrder { get; set; }
    public bool IsAggregationTarget { get; set; }
    public string? TaxCode { get; set; }
    public decimal Balance { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
}

File: app/rust/accounting-system/src/domain/account.rs
use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use utoipa::ToSchema;

/// 勘定科目エンティティ
/// データベースの日本語カラム名と英語プロパティ名をマッピング
#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
pub struct Account {
    #[sqlx(rename = "勘定科目ID")]
    pub account_id: Option<i32>,
    #[sqlx(rename = "勘定科目コード")]
    pub account_code: String,
    #[sqlx(rename = "勘定科目名")]
    pub account_name: String,
    #[sqlx(rename = "勘定科目カナ")]
    pub account_name_kana: Option<String>,
    #[sqlx(rename = "勘定科目種別")]
    pub account_type: String,
    #[sqlx(rename = "合計科目")]
    pub is_summary_account: bool,
    #[sqlx(rename = "BSPL区分")]
    pub bspl_type: Option<String>,
    #[sqlx(rename = "取引要素区分")]
    pub transaction_element_type: Option<String>,
    #[sqlx(rename = "費用区分")]
    pub expense_type: Option<String>,
    #[sqlx(rename = "表示順序")]
    pub display_order: i32,
    #[sqlx(rename = "集計対象")]
    pub is_aggregation_target: bool,
    #[sqlx(rename = "課税取引コード")]
    pub tax_code: Option<String>,
    #[sqlx(rename = "残高")]
    pub balance: Decimal,
    #[sqlx(rename = "作成日時")]
    pub created_at: Option<DateTime<Utc>>,
    #[sqlx(rename = "更新日時")]
    pub updated_at: Option<DateTime<Utc>>,
}

impl Account {
    /// 新規勘定科目を作成
    pub fn new(
        account_code: String,
        account_name: String,
        account_type: String,
        is_summary_account: bool,
    ) -> Self {
        Self {
            account_id: None,
            account_code,
            account_name,
            account_name_kana: None,
            account_type,
            is_summary_account,
            bspl_type: None,
            transaction_element_type: None,
            expense_type: None,
            display_order: 0,
            is_aggregation_target: true,
            tax_code: None,
            balance: Decimal::ZERO,
            created_at: None,
            updated_at: None,
        }
    }
}
